// @tact-lang/compiler/dist/generator/writers/writeExpression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeExpression = exports.writeValue = void 0;
const ast_1 = require("../../grammar/ast");
const resolveExpression_1 = require("../../types/resolveExpression");
const resolveDescriptors_1 = require("../../types/resolveDescriptors");
const types_1 = require("../../types/types");
const resolveFuncTypeUnpack_1 = require("./resolveFuncTypeUnpack");
const map_1 = require("../../abi/map");
const global_1 = require("../../abi/global");
const resolveStrings_1 = require("../../types/resolveStrings");
const id_1 = require("./id");
const struct_1 = require("../../abi/struct");
const resolveFuncType_1 = require("./resolveFuncType");
function isNull(f) {
    if (f.kind === 'null') {
        return true;
    }
    return false;
}
function tryExtractPath(f) {
    if (f.kind === 'id') {
        return [f.value];
    }
    if (f.kind === 'op_field') {
        let p = tryExtractPath(f.src);
        if (p) {
            return [...p, f.name];
        }
        else {
            return null;
        }
    }
    return null;
}
function writeStructConstructor(type, args, ctx) {
    // Check for duplicates
    let name = `__gen_constructor_${type.name}$${args.join('_')}`;
    let renderKey = '$constructor$' + type.name + '$' + args.join(',');
    if (ctx.isRendered(renderKey)) {
        return name;
    }
    ctx.markRendered(renderKey);
    // Generate constructor
    ctx.fun(name, () => {
        ctx.append(`(${(0, resolveFuncType_1.resolveFuncType)(type, ctx)}) ${name}(${args.map((v) => (0, resolveFuncType_1.resolveFuncType)(type.fields.find((v2) => v2.name === v).type, ctx) + ' ' + v).join(', ')}) inline {`);
        ctx.inIndent(() => {
            // Create expressions
            let expressions = type.fields.map((v) => {
                let arg = args.find((v2) => v2 === v.name);
                if (arg) {
                    return arg;
                }
                else if (v.default !== undefined) {
                    return writeValue(v.default, ctx);
                }
                else {
                    throw Error(`Missing argument for field "${v.name}" in struct "${type.name}"`); // Must not happen
                }
            }, ctx);
            ctx.append(`return (${expressions.join(', ')});`);
        });
        ctx.append(`}`);
    });
    return name;
}
function writeValue(s, ctx) {
    if (typeof s === 'bigint') {
        return s.toString(10);
    }
    if (typeof s === 'string') {
        let id = (0, resolveStrings_1.getStringId)(s, ctx.ctx);
        ctx.used(`__gen_str_${id}`);
        return `__gen_str_${id}()`;
    }
    if (typeof s === 'boolean') {
        return s ? 'true' : 'false';
    }
    if (s === null) {
        return 'null()';
    }
    throw Error('Invalid value');
}
exports.writeValue = writeValue;
function writeExpression(f, ctx) {
    //
    // Boolean
    //
    if (f.kind === 'boolean') {
        return f.value ? 'true' : 'false';
    }
    //
    // Number
    //
    if (f.kind === 'number') {
        return f.value.toString(10);
    }
    //
    // String literal
    //
    if (f.kind === 'string') {
        let id = (0, resolveStrings_1.getStringId)(f.value, ctx.ctx);
        ctx.used(`__gen_str_${id}`);
        return `__gen_str_${id}()`;
    }
    //
    // Null
    //
    if (f.kind === 'null') {
        return 'null()';
    }
    //
    // ID Reference
    //
    if (f.kind === 'id') {
        let t = (0, resolveExpression_1.getExpType)(ctx.ctx, f);
        // Handle packed type
        if (t.kind === 'ref') {
            let tt = (0, resolveDescriptors_1.getType)(ctx.ctx, t.name);
            if (tt.kind === 'contract' || tt.kind === 'struct') {
                return (0, resolveFuncTypeUnpack_1.resolveFuncTypeUnpack)(t, (0, id_1.id)(f.value), ctx);
            }
        }
        // Handle constant
        if ((0, resolveDescriptors_1.hasStaticConstant)(ctx.ctx, f.value)) {
            let c = (0, resolveDescriptors_1.getStaticConstant)(ctx.ctx, f.value);
            return writeValue(c.value, ctx);
        }
        return (0, id_1.id)(f.value);
    }
    //
    // Binary operations: *, /, +, -, ||, &&, ==, !=, <, <=, >, >=
    // NOTE: We always wrap in parenthesis to avoid operator precedence issues
    //
    if (f.kind === 'op_binary') {
        // Special case for non-integer types and nullable
        if (f.op === '==' || f.op === '!=') {
            if (isNull(f.left) && isNull(f.right)) {
                if (f.op === '==') {
                    return 'true';
                }
                else {
                    return 'false';
                }
            }
            else if (isNull(f.left) && !isNull(f.right)) {
                if (f.op === '==') {
                    return `null?(${writeExpression(f.right, ctx)})`;
                }
                else {
                    return `(~ null?(${writeExpression(f.right, ctx)}))`;
                }
            }
            else if (!isNull(f.left) && isNull(f.right)) {
                if (f.op === '==') {
                    return `null?(${writeExpression(f.left, ctx)})`;
                }
                else {
                    return `(~ null?(${writeExpression(f.left, ctx)}))`;
                }
            }
        }
        // Special case for address
        let lt = (0, resolveExpression_1.getExpType)(ctx.ctx, f.left);
        let rt = (0, resolveExpression_1.getExpType)(ctx.ctx, f.right);
        if (lt.kind === 'ref' && rt.kind === 'ref' && lt.name === 'Address' && rt.name === 'Address') {
            if (f.op === '==') {
                ctx.used(`__tact_address_eq`);
                return (`__tact_address_eq(${writeExpression(f.left, ctx)}, ${writeExpression(f.right, ctx)})`);
            }
            else if (f.op === '!=') {
                ctx.used(`__tact_address_neq`);
                return (`__tact_address_neq(${writeExpression(f.left, ctx)}, ${writeExpression(f.right, ctx)})`);
            }
            else {
                (0, ast_1.throwError)('Cannot use ' + f.op + ' on addresses', f.ref);
            }
        }
        if (lt.kind === 'map' && rt.kind === 'map') {
            if (f.op === '==') {
                ctx.used(`__tact_cell_eq`);
                return (`__tact_cell_eq(${writeExpression(f.left, ctx)}, ${writeExpression(f.right, ctx)})`);
            }
            else if (f.op === '!=') {
                ctx.used(`__tact_cell_eq`);
                return (`__tact_cell_neq(${writeExpression(f.left, ctx)}, ${writeExpression(f.right, ctx)})`);
            }
            else {
                (0, ast_1.throwError)('Cannot use ' + f.op + ' on maps', f.ref);
            }
        }
        let op;
        if (f.op === '*') {
            op = '*';
        }
        else if (f.op === '/') {
            op = '/';
        }
        else if (f.op === '%') {
            op = '%';
        }
        else if (f.op === '+') {
            op = '+';
        }
        else if (f.op === '-') {
            op = '-';
        }
        else if (f.op === '==') {
            op = '==';
        }
        else if (f.op === '!=') {
            op = '!=';
        }
        else if (f.op === '<') {
            op = '<';
        }
        else if (f.op === '<=') {
            op = '<=';
        }
        else if (f.op === '>') {
            op = '>';
        }
        else if (f.op === '>=') {
            op = '>=';
        }
        else if (f.op === '||') {
            op = '|';
        }
        else if (f.op === '&&') {
            op = '&';
        }
        else if (f.op === '<<') {
            op = '<<';
        }
        else if (f.op === '>>') {
            op = '>>';
        }
        else {
            (0, ast_1.throwError)('Unknown binary operator: ' + f.op, f.ref);
        }
        return '(' + writeExpression(f.left, ctx) + ' ' + op + ' ' + writeExpression(f.right, ctx) + ')';
    }
    //
    // Unary operations: !, -, +, !!
    // NOTE: We always wrap in parenthesis to avoid operator precedence issues
    //
    if (f.kind === 'op_unary') {
        // NOTE: Logical not is written as a bitwise not
        if (f.op === '!') {
            return '(~ ' + writeExpression(f.right, ctx) + ')';
        }
        if (f.op === '-') {
            return '(- ' + writeExpression(f.right, ctx) + ')';
        }
        if (f.op === '+') {
            return '(+ ' + writeExpression(f.right, ctx) + ')';
        }
        // NOTE: Assert function that ensures that the value is not null
        if (f.op === '!!') {
            let t = (0, resolveExpression_1.getExpType)(ctx.ctx, f.right);
            if (t.kind === 'ref') {
                let tt = (0, resolveDescriptors_1.getType)(ctx.ctx, t.name);
                if (tt.kind === 'struct') {
                    ctx.used(`__gen_${tt.name}_not_null`);
                    return `__gen_${tt.name}_not_null(${writeExpression(f.right, ctx)})`;
                }
            }
            ctx.used('__tact_not_null');
            return '__tact_not_null(' + writeExpression(f.right, ctx) + ')';
        }
        (0, ast_1.throwError)('Unknown unary operator: ' + f.op, f.ref);
    }
    //
    // Field Access
    // NOTE: this branch resolves "a.b", where "a" is an expression and "b" is a field name
    //
    if (f.kind === 'op_field') {
        // Resolve the type of the expression
        let src = (0, resolveExpression_1.getExpType)(ctx.ctx, f.src);
        if (src === null || src.kind !== 'ref' || src.optional) {
            (0, ast_1.throwError)(`Cannot access field of non-struct type: ${(0, types_1.printTypeRef)(src)}`, f.ref);
        }
        let srcT = (0, resolveDescriptors_1.getType)(ctx.ctx, src.name);
        // Resolve field
        let field = srcT.fields.find((v) => v.name === f.name);
        let cst = srcT.constants.find((v) => v.name === f.name);
        if (!field && !cst) {
            (0, ast_1.throwError)(`Cannot find field "${f.name}" in struct "${srcT.name}"`, f.ref);
        }
        if (field) {
            // Trying to resolve field as a path
            let path = tryExtractPath(f);
            if (path) {
                // Prepare path
                let convertedPath = [];
                convertedPath.push((0, id_1.id)(path[0]));
                convertedPath.push(...path.slice(1));
                let idd = convertedPath.join(`'`);
                // Special case for structs
                if (field.type.kind === 'ref') {
                    let ft = (0, resolveDescriptors_1.getType)(ctx.ctx, field.type.name);
                    if (ft.kind === 'struct' || ft.kind === 'contract') {
                        return (0, resolveFuncTypeUnpack_1.resolveFuncTypeUnpack)(field.type, idd, ctx);
                    }
                }
                return idd;
            }
            // Getter instead of direct field access
            ctx.used(`__gen_${srcT.name}_get_${field.name}`);
            return `__gen_${srcT.name}_get_${field.name}(${writeExpression(f.src, ctx)})`;
        }
        else {
            return writeValue(cst.value, ctx);
        }
    }
    //
    // Static Function Call
    //
    if (f.kind === 'op_static_call') {
        // Check global functions
        if (global_1.GlobalFunctions[f.name]) {
            return global_1.GlobalFunctions[f.name].generate(ctx, f.args.map((v) => (0, resolveExpression_1.getExpType)(ctx.ctx, v)), f.args, f.ref);
        }
        let sf = (0, resolveDescriptors_1.getStaticFunction)(ctx.ctx, f.name);
        let n = f.name;
        if (sf.ast.kind === 'def_native_function') {
            n = sf.ast.nativeName;
            if (n.startsWith('__tact')) {
                ctx.used(n);
            }
        }
        else {
            ctx.used(n);
            n = (0, id_1.fn)(n);
        }
        return n + '(' + f.args.map((a) => writeExpression(a, ctx)).join(', ') + ')';
    }
    //
    // Struct Constructor
    //
    if (f.kind === 'op_new') {
        let src = (0, resolveDescriptors_1.getType)(ctx.ctx, f.type);
        // Write a constructor
        let id = writeStructConstructor(src, f.args.map((v) => v.name), ctx);
        ctx.used(id);
        // Write an expression
        let expressions = f.args.map((v) => writeExpression(v.exp, ctx), ctx);
        return `${id}(${expressions.join(', ')})`;
    }
    //
    // Object-based function call
    //
    if (f.kind === 'op_call') {
        // Resolve source type
        let src = (0, resolveExpression_1.getExpType)(ctx.ctx, f.src);
        if (src === null) {
            (0, ast_1.throwError)(`Cannot call function of non - direct type: ${(0, types_1.printTypeRef)(src)} `, f.ref);
        }
        // Reference type
        if (src.kind === 'ref') {
            if (src.optional) {
                (0, ast_1.throwError)(`Cannot call function of non - direct type: ${(0, types_1.printTypeRef)(src)} `, f.ref);
            }
            // Render function call
            let t = (0, resolveDescriptors_1.getType)(ctx.ctx, src.name);
            // Check struct ABI
            if (t.kind === 'struct') {
                let abi = struct_1.StructFunctions[f.name];
                if (abi) {
                    return abi.generate(ctx, [src, ...f.args.map((v) => (0, resolveExpression_1.getExpType)(ctx.ctx, v))], [f.src, ...f.args], f.ref);
                }
            }
            // Resolve function
            let ff = t.functions.get(f.name);
            let name = `__gen_${src.name}_${f.name}`;
            if (ff.ast.kind === 'def_function') {
                ctx.used(name);
                name = (0, id_1.fn)(name);
            }
            else {
                name = ff.ast.nativeName;
                if (name.startsWith('__tact')) {
                    ctx.used(name);
                }
            }
            // Render
            if (ff.isMutating) {
                let s = writeExpression(f.src, ctx);
                if (f.args.length === 1) {
                    let t = (0, resolveExpression_1.getExpType)(ctx.ctx, f.args[0]);
                    if (t.kind === 'ref') {
                        let tt = (0, resolveDescriptors_1.getType)(ctx.ctx, t.name);
                        if (tt.kind === 'contract' || tt.kind === 'struct') {
                            ctx.used(`__gen_${tt.name}_unpack`);
                            return `${s}~${name}(__gen_${tt.name}_unpack(${writeExpression(f.args[0], ctx)}))`;
                        }
                    }
                }
                return `${s}~${name}(${[...f.args.map((a) => writeExpression(a, ctx))].join(', ')})`;
            }
            else {
                let s = writeExpression(f.src, ctx);
                return `${name}(${[s, ...f.args.map((a) => writeExpression(a, ctx))].join(', ')})`;
            }
        }
        // Map types
        if (src.kind === 'map') {
            let abf = map_1.MapFunctions[f.name];
            if (!abf) {
                (0, ast_1.throwError)(`Map function "${f.name}" not found`, f.ref);
            }
            return abf.generate(ctx, [src, ...f.args.map((v) => (0, resolveExpression_1.getExpType)(ctx.ctx, v))], [f.src, ...f.args], f.ref);
        }
        (0, ast_1.throwError)(`Cannot call function of non - direct type: ${(0, types_1.printTypeRef)(src)} `, f.ref);
    }
    //
    // Init of
    //
    if (f.kind === 'init_of') {
        ctx.used(`__gen_${f.name}_init_child`);
        ctx.used(`__tact_context`);
        return `${(0, id_1.fn)(`__gen_${f.name}_init_child`)}(${['__tact_context_sys', ...f.args.map((a) => writeExpression(a, ctx))].join(', ')})`;
    }
    //
    // Unreachable
    //
    throw Error('Unknown expression');
}
exports.writeExpression = writeExpression;
